packages/circuits/jwt-verifier-template.circom

```diff
@@ -0,0 +1,327 @@
+
+    // Calculate SHA256 hash of the JWT message
+    signal sha[256] <== Sha256Bytes(maxMessageLength)(message, messageLength);
+
+    // Pack SHA output bytes to int[] for RSA input message
+    var rsaMessageSize = (256 + n) \ n; // Adjust based on RSA chunk size
+    component rsaMessage[rsaMessageSize];
+    for (var i = 0; i < rsaMessageSize; i++) {
+        rsaMessage[i] = Bits2Num(n);
+    }
+    for (var i = 0; i < 256; i++) {
+        rsaMessage[i \ n].in[i % n] <== sha[255 - i];
+    }
+    for (var i = 256; i < n * rsaMessageSize; i++) {
+        rsaMessage[i \ n].in[i % n] <== 0;
+    }
+
+    // Verify RSA signature
+    component rsaVerifier = RSAVerifier65537(n, k);
+    for (var i = 0; i < rsaMessageSize; i++) {
+        rsaVerifier.message[i] <== rsaMessage[i].out;
+    }
+    for (var i = rsaMessageSize; i < k; i++) {
+        rsaVerifier.message[i] <== 0;
+    }
+    rsaVerifier.modulus <== pubkey;
+    rsaVerifier.signature <== signature;
+
+    // Calculate the pubkey hash
+    publicKeyHash <== PoseidonLarge(n, k)(pubkey);
+
+    /// Calculate the JWT nullifier
+    var k2ChunkedSize = k >> 1;
+    if(k % 2 == 1) {
+        k2ChunkedSize += 1;
+    }
+    signal signHash;
+    signal signInts[k2ChunkedSize];
+    (signHash, signInts) <== HashSign(n,k)(signature);
+    jwtNullifier <== Poseidon(1)([signHash]);
+
+    // Assert that period exists at periodIndex
+    signal period <== ItemAtIndex(maxMessageLength)(message, periodIndex);
+    period === 46;
+
+    // Assert that period is unique
+    signal periodCount <== CountCharOccurrences(maxMessageLength)(message, 46);
+    periodCount === 1;
+
+    // Find the real message length
+    signal realMessageLength <== FindRealMessageLength(maxMessageLength)(message);
+
+    // Calculate the length of the Base64 encoded header and payload
+    signal b64HeaderLength <== periodIndex;
+    signal b64PayloadLength <== realMessageLength - b64HeaderLength - 1;
+
+    // Extract the Base64 encoded header and payload from the message
+    signal b64Header[maxB64HeaderLength] <== SelectSubArrayBase64(maxMessageLength, maxB64HeaderLength)(message, 0, b64HeaderLength);
+    signal b64Payload[maxB64PayloadLength] <== SelectSubArrayBase64(maxMessageLength, maxB64PayloadLength)(message, b64HeaderLength + 1, b64PayloadLength);
+
+    // Calculate the maximum length of the decoded header and payload
+    var maxHeaderLength = (maxB64HeaderLength * 3) \ 4;
+    var maxPayloadLength = (maxB64PayloadLength * 3) \ 4;
+
+    // Decode the Base64 encoded header and payload
+    signal header[maxHeaderLength] <== Base64Decode(maxHeaderLength)(b64Header);
+    signal payload[maxPayloadLength] <== Base64Decode(maxPayloadLength)(b64Payload);
+
+    // Verify if the typ in the header is "JWT"
+    var typLength = JWT_TYP_LENGTH();
+    var typ[typLength] = JWT_TYP();
+    signal typMatch[typLength] <== RevealSubstring(maxHeaderLength, typLength, 0)(header, jwtTypStartIndex, typLength);
```

Comment:
Could you share the context about disabling the `shouldCheckUniqueness` here?
Also, I don't understand how `RevealSubstring` prevents unrelated `jwtTypStartIndex` such that the substring from `jwtTypStartIndex` contains "JWT" but is not for "typ"?
(The zk-regex prevents it by masking all characters except for the value of "typ" defined by the regex.)

Reply:
We verify if the key `typ` is unique by using the start index, then we calculate the start index for the value which is `JWT` which we need to check if it's unique since the key is already unique.

Reply:
> we calculate the start index for the value which is JWT which we need to check if it's unique since the key is already unique.

Sorry, could you share concrete lines of constraints for this validation?

Reply:
@SoraSuegami Sorry, I had missed this check. `JWT_TYP()` gives the string - `"typ":"JWT"` which we should check is unique. Thanks for catching this, I've fixed it in the latest commit.